src/Microsoft.ML.Data/DataLoadSave/TransformerChain.cs

```diff
@@ -0,0 +1,227 @@
+        {
+            var xfs = new List<ITransformer>();
+            var scopes = new List<TransformerScope>();
+            for (int i = 0; i < _transformers.Length; i++)
+            {
+                if ((_scopes[i] & scopeFilter) != TransformerScope.None)
+                {
+                    xfs.Add(_transformers[i]);
+                    scopes.Add(_scopes[i]);
+                }
+            }
+            return new TransformerChain<ITransformer>(xfs.ToArray(), scopes.ToArray());
+        }
+
+        public TransformerChain<TNewLast> Append<TNewLast>(TNewLast transformer, TransformerScope scope = TransformerScope.Everything)
+            where TNewLast : class, ITransformer
+        {
+            Contracts.CheckValue(transformer, nameof(transformer));
+            return new TransformerChain<TNewLast>(_transformers.Append(transformer).ToArray(), _scopes.Append(scope).ToArray());
+        }
+
+        public void Save(ModelSaveContext ctx)
+        {
+            ctx.CheckAtModel();
+            ctx.SetVersionInfo(GetVersionInfo());
+
+            ctx.Writer.Write(_transformers.Length);
+
+            for (int i = 0; i < _transformers.Length; i++)
+            {
+                ctx.Writer.Write((int)_scopes[i]);
+                var dirName = string.Format(TransformDirTemplate, i);
+                ctx.SaveModel(_transformers[i], dirName);
+            }
+        }
+
+        /// <summary>
+        /// The loading constructor of transformer chain. Reverse of <see cref="Save(ModelSaveContext)"/>.
+        /// </summary>
+        internal TransformerChain(IHostEnvironment env, ModelLoadContext ctx)
+        {
+            int len = ctx.Reader.ReadInt32();
+            _transformers = new ITransformer[len];
+            _scopes = new TransformerScope[len];
+            for (int i = 0; i < len; i++)
+            {
+                _scopes[i] = (TransformerScope)(ctx.Reader.ReadInt32());
+                var dirName = string.Format(TransformDirTemplate, i);
+                ctx.LoadModel<ITransformer, SignatureLoadModel>(env, out _transformers[i], dirName);
+            }
+            if (len > 0)
+                LastTransformer = _transformers[len - 1] as TLastTransformer;
+            else
+                LastTransformer = null;
+        }
+
+        public void SaveTo(IHostEnvironment env, Stream outputStream)
+        {
+            using (var ch = env.Start("Saving pipeline"))
+            {
+                using (var rep = RepositoryWriter.CreateNew(outputStream, ch))
+                {
+                    ch.Trace("Saving transformer chain");
+                    ModelSaveContext.SaveModel(rep, this, TransformerChain.LoaderSignature);
+                    rep.Commit();
+                }
+            }
+        }
+
+        public IEnumerator<ITransformer> GetEnumerator() => ((IEnumerable<ITransformer>)_transformers).GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
```

Comment:
>IEnumerator IEnumerable.GetEnumerator() => GetEnumerator(); [](start = 8, length = 59)

can you add implicit private?

Reply:
Actually it's not private. Nor is it public of protected :)

---
In reply to: [211688343](https://github.com/dotnet/machinelearning/pull/688#discussion_r211688343) [](ancestors = 211688343)